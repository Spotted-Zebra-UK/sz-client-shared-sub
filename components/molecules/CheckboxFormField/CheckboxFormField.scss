@import '../../../variables';

.CheckboxFormField {
  height: 46px;

  svg {
    margin-right: 5px;
  }

  &__Label {
    cursor: pointer;

    &__Top {
      display: flex;
      align-items: center;

      .Checkbox {
        position: absolute;
        opacity: 0;
        height: 0;
        width: 0;
      }
    }

    &__Error {
      color: $purple;
      font-size: $p2-font-size;
      margin-top: 8px;
    }
  }
}
