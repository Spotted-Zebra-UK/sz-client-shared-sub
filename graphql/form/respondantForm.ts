import { gql } from '@apollo/client';

// TODO: Refactor, use generated query instead of this.
export const RESPONDANT_FORM_QUERY = gql`
  query RespondantFormCompletedCheck(
    $associatedId: Int!
    $formOwnerId: Int!
    $formType: FormType!
    $respondantId: Int
  ) {
    respondantForm: RespondantForm(
      associatedId: $associatedId
      formOwnerId: $formOwnerId
      formType: $formType
      respondantId: $respondantId
    ) {
      isComplete
      submittedFormLogId
      formType
      formOwnerId
      associatedId
      formId
      respondantFieldWithAnswers {
        field {
          id
          question
          hint
          settings {
            allowFreeText
            searchable
          }
          isArchived
          fieldType
          ... on CompanyEmployeeSelectField {
            dynamicSelectOptions
          }
          ... on MultiSelectField {
            selectOptions
          }
          ... on SingleSelectField {
            selectOptions
          }
        }
        displayOrder
        isOptional
        answer
        selectOptionAnswerIsFreeText
      }
    }
  }
`;

// TODO: Refactor, use generated query instead of this.
export const RESPONDANT_FORM_UPDATE_MUTATION = gql`
  mutation RespondantFormUpdate(
    $associatedId: Int!
    $formOwnerId: Int!
    $formType: FormType!
    $answers: [UpdateRespondantFormAnswerArgs!]
  ) {
    respondantForm: RespondantFormUpdate(
      associatedId: $associatedId
      formOwnerId: $formOwnerId
      formType: $formType
      answers: $answers
    ) {
      isComplete
      submittedFormLogId
      formType
      formOwnerId
      associatedId
      formId
      respondantFieldWithAnswers {
        field {
          id
          question
          hint
          settings {
            allowFreeText
            searchable
          }
          isArchived
          fieldType
          ... on CompanyEmployeeSelectField {
            dynamicSelectOptions
          }
          ... on MultiSelectField {
            selectOptions
          }
          ... on SingleSelectField {
            selectOptions
          }
        }
        displayOrder
        isOptional
        answer
        selectOptionAnswerIsFreeText
      }
    }
  }
`;
