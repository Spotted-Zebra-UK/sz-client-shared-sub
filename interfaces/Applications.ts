export enum Application {
  CANDIDATE = 'candidate',
  COMPANY = 'company',
}
