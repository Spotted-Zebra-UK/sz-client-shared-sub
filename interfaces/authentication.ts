export interface IRequestTokenRefreshResponse {
  requestTokenRefresh: {
    accessToken: string;
    refreshToken: string;
  };
}

export interface IRequestTokenRefreshInput {
  accessToken: string;
  refreshToken: string;
}

export interface IDeauthenticateInput {
  accessToken: string;
}
