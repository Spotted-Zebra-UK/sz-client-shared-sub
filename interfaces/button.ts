export type TButtonType = 'submit' | 'button' | 'link';
