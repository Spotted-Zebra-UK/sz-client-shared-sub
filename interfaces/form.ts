export type TTextFromFieldValue = string;
export type TSelectFormFieldOptionValue = string | number | undefined;
export type TSelectFormFieldValue =
  | string
  | number
  | undefined
  | string[]
  | number[];
export type TCheckboxFormFieldValue = boolean;
export type TFormFieldValue =
  | TSelectFormFieldValue
  | TTextFromFieldValue
  | TCheckboxFormFieldValue;
