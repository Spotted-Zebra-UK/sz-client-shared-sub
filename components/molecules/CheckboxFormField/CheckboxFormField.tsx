import './CheckboxFormField.scss';
import React, { FC } from 'react';
import { ReactComponent as CheckboxCheckedIcon } from '../../../icons/CheckboxChecked.svg';
import { ReactComponent as CheckboxEmptyIcon } from '../../../icons/CheckboxEmpty.svg';
import { TCheckboxFormFieldValue } from '../../../interfaces/form';
import Checkbox from '../../atoms/Checkbox/Checkbox';

interface ICheckboxFormField {
  className?: string;
  isDisabled?: boolean;
  name: string;
  id: string;
  onChange: (value: TCheckboxFormFieldValue, name: string) => void;
  value: boolean;
  label?: string | React.ReactNode;
  error?: string;
}

const CheckboxFormField: FC<ICheckboxFormField> = ({
  value,
  className,
  label,
  id,
  error,
  ...restProps
}) => {
  return (
    <div className={`CheckboxFormField${className ? ` ${className}` : ''}`}>
      <label htmlFor={id} className="CheckboxFormField__Label">
        <div className="CheckboxFormField__Label__Top">
          <Checkbox value={value} id={id} {...restProps} />
          {value ? <CheckboxCheckedIcon /> : <CheckboxEmptyIcon />}
          {label ? (
            <div className="CheckboxFormField__Label__Top__Input">{label}</div>
          ) : null}
        </div>
        <div className="CheckboxFormField__Label__Error">{error}</div>
      </label>
    </div>
  );
};

export default CheckboxFormField;
