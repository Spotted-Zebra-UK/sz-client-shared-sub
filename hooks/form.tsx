import { useState } from 'react';
import validate from '../helpers/validate';
import { TFormFieldValue } from '../interfaces/form';

export type TFormValues = {
  [key in string]: TFormFieldValue;
};

export type TFormErrors = {
  [key in string]: string;
};

interface IUseFormOptions {
  initialValues: TFormValues;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  validationShema?: any;
}

interface IUseFormProps {
  onSubmit: (values: TFormValues) => void;
  options: IUseFormOptions;
}

export const useForm = ({
  options,
  onSubmit,
}: IUseFormProps): [
  TFormValues,
  TFormErrors | undefined,
  (newValue: TFormFieldValue, name: string) => void,
  (
    event:
      | React.FormEvent<HTMLFormElement>
      | React.MouseEvent<HTMLButtonElement, MouseEvent>
      | undefined
  ) => void
] => {
  const [values, setValues] = useState<TFormValues>(options.initialValues);
  const [errors, setErrors] = useState<TFormErrors | undefined>();

  const validateForm = () => {
    if (!options.validationShema) {
      return undefined;
    }

    return validate(values, options.validationShema);
  };

  const handleChange = (newValue: TFormFieldValue, name: string) => {
    setValues(prevValues => ({ ...prevValues, [name]: newValue }));

    if (errors && errors[name]) {
      setErrors(prevErrors => {
        return { ...prevErrors, [name]: '' };
      });
    }
  };

  const handleSubmit = (
    event:
      | React.FormEvent<HTMLFormElement>
      | React.MouseEvent<HTMLButtonElement, MouseEvent>
      | undefined
  ) => {
    event?.preventDefault();
    const newErrors = validateForm();
    if (!newErrors) {
      onSubmit(values);
    } else {
      setErrors(newErrors);
    }
  };

  return [values, errors, handleChange, handleSubmit];
};
