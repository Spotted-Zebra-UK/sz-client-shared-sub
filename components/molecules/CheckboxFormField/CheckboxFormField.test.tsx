import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { ReactComponent as CheckboxCheckedIcon } from '../../../icons/CheckboxChecked.svg';
import { ReactComponent as CheckboxEmptyIcon } from '../../../icons/CheckboxEmpty.svg';
import { TCheckboxFormFieldValue } from '../../../interfaces/form';
import CheckboxFormField from './CheckboxFormField';

describe('CheckboxFormField', () => {
  let onChange: (value: TCheckboxFormFieldValue, name: string) => void;
  const name = 'test-name';
  const id = 'test-id';

  beforeEach(() => {
    onChange = jest.fn();
  });

  it('should render correct with required props when value if false', () => {
    const wrapper = shallow(
      <CheckboxFormField
        onChange={onChange}
        name={name}
        id={id}
        value={false}
      />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(CheckboxEmptyIcon)).toHaveLength(1);
    expect(wrapper.find(CheckboxCheckedIcon)).toHaveLength(0);
  });

  it('should render correct when value is true', () => {
    const wrapper = shallow(
      <CheckboxFormField onChange={onChange} name={name} id={id} value />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(CheckboxEmptyIcon)).toHaveLength(0);
    expect(wrapper.find(CheckboxCheckedIcon)).toHaveLength(1);
  });

  it('should render correctly when label provided', () => {
    const label = 'test-label';
    const wrapper = shallow(
      <CheckboxFormField
        label={label}
        onChange={onChange}
        name={name}
        id={id}
        value
      />
    );

    expect(wrapper.find('.CheckboxFormField__Label__Top__Input')).toHaveLength(
      1
    );
    expect(
      wrapper.find('.CheckboxFormField__Label__Top__Input').text()
    ).toEqual(label);
  });

  it('should render correctly when className provided', () => {
    const className = 'test-className';
    const wrapper = shallow(
      <CheckboxFormField
        className={className}
        onChange={onChange}
        name={name}
        id={id}
        value
      />
    );

    expect(wrapper.hasClass('CheckboxFormField'));
    expect(wrapper.hasClass(className)).toBe(true);
  });

  it('should render correctly when error provided', () => {
    const error = 'test-error';
    const wrapper = shallow(
      <CheckboxFormField
        error={error}
        onChange={onChange}
        name={name}
        id={id}
        value
      />
    );

    expect(wrapper.find('.CheckboxFormField__Label__Error').text()).toEqual(
      error
    );
  });

  it('should call onChange on input change', () => {
    const wrapper = mount(
      <CheckboxFormField onChange={onChange} name={name} id={id} value />
    );

    wrapper.find('input').simulate('change');

    expect(onChange).toHaveBeenCalledWith(false, name);
  });
});
