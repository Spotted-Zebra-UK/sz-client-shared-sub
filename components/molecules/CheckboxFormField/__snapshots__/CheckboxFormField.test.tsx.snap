// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxFormField should render correct when value is true 1`] = `
<div
  className="CheckboxFormField"
>
  <label
    className="CheckboxFormField__Label"
    htmlFor="test-id"
  >
    <div
      className="CheckboxFormField__Label__Top"
    >
      <Checkbox
        id="test-id"
        name="test-name"
        onChange={[MockFunction]}
        value={true}
      />
      <ForwardRef(SvgCheckboxChecked) />
    </div>
    <div
      className="CheckboxFormField__Label__Error"
    />
  </label>
</div>
`;

exports[`CheckboxFormField should render correct with required props when value if false 1`] = `
<div
  className="CheckboxFormField"
>
  <label
    className="CheckboxFormField__Label"
    htmlFor="test-id"
  >
    <div
      className="CheckboxFormField__Label__Top"
    >
      <Checkbox
        id="test-id"
        name="test-name"
        onChange={[MockFunction]}
        value={false}
      />
      <ForwardRef(SvgCheckboxEmpty) />
    </div>
    <div
      className="CheckboxFormField__Label__Error"
    />
  </label>
</div>
`;
